gather.go

```diff
@@ -80,6 +81,7 @@ func (a *Agent) GatherCandidates() error {
 }
 
 func (a *Agent) gatherCandidates(ctx context.Context) {
+	defer close(a.gatherCandidateDone)
 	if err := a.setGatheringState(GatheringStateGathering); err != nil {
 		a.log.Warnf("failed to set gatheringState to GatheringStateGathering: %v", err)
 		return
```

Comment:
for the sake of leaks, I think you may want to change those .Done's to be `defer wg.Done()` at the beginning of each of the goroutines, to protect against any panics preventing Closing.